src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,103 @@
 package core.basesyntax;
 
 public class MyHashMap<K, V> implements MyMap<K, V> {
+    private static final int DEFAULT_CAPACITY = 16;
+    private static final double LOAD_FACTOR = 0.75;
+    private static final int RESIZE_INDEX = 2;
+    private int size;
+    private int capacity;
+    private Node<K, V>[] buckets;
+
+    public MyHashMap() {
+        this.capacity = DEFAULT_CAPACITY;
+        this.buckets = new Node[capacity];
+    }
 
     @Override
     public void put(K key, V value) {
+        int hash = hashCode(key);
+        if (buckets[hash] == null) {
+            buckets[hash] = new Node<>(hash, key, value, null);
+            size++;
+        } else {
+            Node<K, V> node = buckets[hash];
+            while (node != null) {
+                if (keyEquals(node.key, key)) {
+                    node.value = value;
+                    return;
+                }
+                node = node.next;
+            }
+            buckets[hash] = new Node<>(hash, key, value, buckets[hash]);
+            size++;
+        }
 
+        if ((double) size / capacity >= LOAD_FACTOR) {
+            resize();
+        }
     }
 
     @Override
     public V getValue(K key) {
+        int hash = hashCode(key);
+        Node<K, V> node = buckets[hash];
+        while (node != null) {
+            if (keyEquals(node.key, key)) {
+                return node.value;
+            }
+            node = node.next;
+        }
         return null;
     }
 
     @Override
     public int getSize() {
-        return 0;
+        return size;
+    }
+
+    private void resize() {
+        capacity = capacity * RESIZE_INDEX;
+        Node<K, V>[] resizeBuckets = new Node[capacity];
+
+        for (Node<K, V> node : buckets) {
+            while (node != null) {
+                int newHashCode = hashCode(node.key) % capacity;
+                if (resizeBuckets[newHashCode] == null) {
+                    resizeBuckets[newHashCode] = new Node<>(
+                            newHashCode, node.key, node.value, null);
+                } else {
+                    Node<K, V> newNode = new Node<>(
+                            newHashCode, node.key, node.value, resizeBuckets[newHashCode]);
+                    resizeBuckets[newHashCode] = newNode;
+                }
+                node = node.next;
+            }
+        }
+        buckets = resizeBuckets;
+    }
+
+    private boolean keyEquals(K k1, K k2) {
+        if (k1 == null) {
+            return k2 == null;
+        }
+        return k1.equals(k2);
+    }
+
+    private int hashCode(K key) {
+        return (key == null) ? 0 : Math.abs(key.hashCode()) % capacity;
+    }
+
+    public static class Node<K, V> {
+        private final K key;
+        private final int hash;
+        private V value;
+        private Node<K, V> next;
+
+        public Node(int hash, K key, V value, Node<K, V> next) {
```

Comment:
```suggestion
        private Node(int hash, K key, V value, Node<K, V> next) {
```
won't be accessible outside anyway, let's not confuse anyone